tokio/src/util/wake_list.rs

```diff
@@ -37,12 +37,34 @@ impl WakeList {
     }
 
     pub(crate) fn wake_all(&mut self) {
+        struct DropGuard {
+            wakers: *mut Waker,
+            start: usize,
+            end: usize,
+        }
+
+        impl Drop for DropGuard {
+            fn drop(&mut self) {
+                let slice = ptr::slice_from_raw_parts_mut(
+                    unsafe { self.wakers.add(self.start) },
+                    self.end - self.start,
+                );
+                // SAFETY: Elements in `start..end` are initialized, so we can drop them.
+                unsafe { ptr::drop_in_place(slice) };
+            }
+        }
+
         assert!(self.curr <= NUM_WAKERS);
-        while self.curr > 0 {
-            self.curr -= 1;
-            // SAFETY: The first `curr` elements of `WakeList` are initialized, so by decrementing
-            // `curr`, we can take ownership of the last item.
-            let waker = unsafe { ptr::read(self.inner[self.curr].as_mut_ptr()) };
+
+        let mut guard = DropGuard {
+            wakers: self.inner.as_mut_ptr().cast::<Waker>(),
+            start: 0,
+            end: mem::take(&mut self.curr),
+        };
+        while guard.start < guard.end {
+            // SAFETY: Elements between `start..end` are initialized.
+            let waker = unsafe { ptr::read(guard.wakers.add(guard.start)) };
+            guard.start += 1;
```

Comment:
The reason why this `ptr::read` operation is safe is that by incrementing `guard.start` by one, the guard gives up ownership of the waker, so we can transfer ownership of it to the local variable.